pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -617,34 +621,81 @@ func (c *Cluster) saveMonConfig() error {
 	d := c.makeDeployment(m, hostname)
-	logger.Debugf("Starting mon: %+v", d.Name)
-	_, err := c.context.Clientset.AppsV1().Deployments(c.Namespace).Create(d)
-	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create mon deployment %s. %+v", m.ResourceName, err)
-		}
-		logger.Infof("deployment for mon %s already exists. updating if needed", m.ResourceName)
+	existingDeployment, err := c.context.Clientset.AppsV1().Deployments(c.Namespace).Get(d.Name, metav1.GetOptions{})
+	if err == nil {
+		deploymentExists = true
+		pvcExists = opspec.DaemonVolumesContainsPVC(existingDeployment.Spec.Template.Spec.Volumes)
+	} else if !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to get mon deployment %s. %+v", d.Name, err)
+	}
+
+	// persistent storage is not altered after the deployment is created. this
+	// means we need to be careful when updating the deployment to avoid new
+	// changes to the crd to change an existing pod's persistent storage. the
+	// deployment spec created above does not specify persistent storage. here
+	// we add in PVC or HostPath storage based on an existing deployment OR on
+	// the current state of the CRD.
+	if pvcExists || (!deploymentExists && c.spec.Mon.VolumeClaimTemplate != nil) {
+		pvcName := makeDeploymentPVCName(m)
+		d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes, opspec.DaemonVolumesDataPVC(pvcName))
+		logger.Debugf("adding pvc volume source %s to mon deployment %s", pvcName, d.Name)
+	} else {
+		d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes, opspec.DaemonVolumesDataHostPath(m.DataPathMap)...)
+		logger.Debugf("adding host path volume source to mon deployment %s", d.Name)
+	}
+
+	if deploymentExists {
+		return c.updateMon(m, d)
+	}
 
-		// Always invoke ceph version before an upgrade so we are sure to be up-to-date
-		daemonType := string(config.MonType)
-		var cephVersionToUse cephver.CephVersion
-		currentCephVersion, err := client.LeastUptodateDaemonVersion(c.context, c.clusterInfo.Name, daemonType)
+	if c.spec.Mon.VolumeClaimTemplate != nil {
+		pvc, err := c.makeDeploymentPVC(m)
 		if err != nil {
-			logger.Warningf("failed to retrieve current ceph %s version. %+v", daemonType, err)
-			logger.Debug("could not detect ceph version during update, this is likely an initial bootstrap, proceeding with c.clusterInfo.CephVersion")
-			cephVersionToUse = c.clusterInfo.CephVersion
-		} else {
-			logger.Debugf("current cluster version for monitors before upgrading is: %+v", currentCephVersion)
-			cephVersionToUse = currentCephVersion
+			return fmt.Errorf("failed to make mon %s pvc. %+v", d.Name, err)
 		}
-		if err := updateDeploymentAndWait(c.context, d, c.Namespace, daemonType, m.DaemonName, cephVersionToUse); err != nil {
-			return fmt.Errorf("failed to update mon deployment %s. %+v", m.ResourceName, err)
+		_, err = c.context.Clientset.CoreV1().PersistentVolumeClaims(c.Namespace).Create(pvc)
+		if err != nil {
```

Comment:
How about checking for the error that it already exists? In that case we don't need to return an error. There is a small chance the PVC has already been created. Perhaps the admin could have already created his own PVCs as well instead of depending on Rook to create them.

Reply:
good point. i didn't consider that admins might create pvcs.